### DISCUSSION TITLE
Why do I get the Must call `setClient()` with an instance of MongoClient error ?

### DISCUSSION THREAD

**Discussion by robots4life** [Automattic/mongoose#11494] [Q&A] [answered]
https://mongoosejs.com/docs/api/connection.html#connection_Connection-setClient

> Set the [MongoDB driver MongoClient](https://mongodb.github.io/node-mongodb-native/3.5/api/MongoClient.html) instance that this connection uses to talk to MongoDB. This is useful if you already have a MongoClient instance, and want to reuse it.

```js
import mongoose from 'mongoose';

import { MongoClient } from 'mongodb';

async function run() {
    try {
        const uri = 'mongodb://localhost:27017';

        // Create a new MongoClient
        const client = new MongoClient(uri);

	console.log(client);

        const conn = mongoose.createConnection().setClient(client);

        conn.getClient(); // MongoClient { ... }
        conn.readyState; // 1, means 'CONNECTED'

    } catch (error) {
        console.log(error);
    }
}
run();
```
This returns `MongooseError: Must call `setClient()` with an instance of MongoClient`. However it logs the client just fine, so I am not sure what I am doing wrong. How do I use `setClient()` correctly please ?

From what I understand is that I can use `setClient` to set the Mongoose client to the client that established a connection with MongoDB with the native driver, is this correct at all or have I got this wrong ? If so, what does `setClient` do exactly ?

**Comment by vkarpov15** [comment, marked as answer]
The error message I get when running the above script is:

```
MongooseError: Cannot call `setClient()` with a MongoClient that you have not called `connect()` on yet.
```

Which is fairly self explanatory, you need to do `await client.connect()` before calling `setClient()`.

The error message you posted, though, makes me think that you have an incompatible version of the mongodb driver. Try running `npm list | grep "mongo"` and paste the output here. The output should look something like:

```
projectname@0.0.2 /home/user/projectname
└─┬ mongoose@6.2.5
  ├─┬ mongodb@4.3.1
  │ ├─┬ mongodb-connection-string-url@2.5.2
```

**Reply by robots4life** [reply]
Thank you for your reply and help.

Unfortunately the error persists with the included `await client.connect()`.

```js
import mongoose from 'mongoose';

import { MongoClient } from 'mongodb';

async function run() {
	try {
		const uri = 'mongodb://localhost:27017';

		// Create a new MongoClient
		const client = new MongoClient(uri);

		await client.connect();

		console.log(client);

		const conn = mongoose.createConnection().setClient(client);

		conn.getClient(); // MongoClient { ... }
		conn.readyState; // 1, means 'CONNECTED'
	} catch (error) {
		console.log(error);
	}
}
run();
```

`npm list | grep "mongo"` returns the following output.

```shell
mongo-mongoose@0.0.1 /shared/httpd/mongo-mongoose
├── mongodb@4.4.1
├── mongoose@6.2.5
```

`npm ls`

```shell
mongo-mongoose@0.0.1 /shared/httpd/mongo-mongoose
├── @sveltejs/adapter-auto@1.0.0-next.31
├── @sveltejs/kit@1.0.0-next.295
├── env-cmd@10.1.0
├── eslint-config-prettier@8.5.0
├── eslint-plugin-svelte3@3.4.1
├── eslint@7.32.0
├── mongodb@4.4.1
├── mongoose@6.2.5
├── prettier-plugin-svelte@2.6.0
├── prettier@2.5.1
└── svelte@3.46.4
```

Here is a simple repo with the above behavior.
https://github.com/robots4life/mongo-mongoose/blob/master/src/routes/index.js

**Reply by vkarpov15** [reply]
Incompatible versions of mongodb and mongoose. Mongoose 6.2 uses mongodb@4.3, so you're trying to pass a `MongoClient` instance from a different version of the MongoDb Node driver

**Reply by robots4life** [reply]
Hmm, so would you kindly mind sharing the correct install commands or what I would have to do so that I can start using Mongoose with MongoDB server version: 4.4.12 ?

This is the MongoDB server version I have. `MongoDB server version: 4.4.12`

When I look here https://mongoosejs.com/docs/compatibility.html I don't see that I should use a lower version of the MongoDB server.

`MongoDB Server 4.4.x: mongoose ^5.10.0 or 6.x`

I have `mongodb@4.4.1` and `mongoose@6.2.5`, so it should work, no ?

I also looked at https://docs.mongodb.com/drivers/node/current/compatibility/ and made sure to have the correct Node version for the corresponding MongoDB NodeJS Driver.

Node.js Driver Version 4.x : Node Version v14.19.0 `nvm use lts/fermium`

I also checked here https://mongodb.github.io/node-mongodb-native/4.4/ to see if I can do something different. `npm install mongodb`

EDIT:
I deleted the `node-modules` folder and `package-lock.json` file and adjusted `package.json` like so.

`package.json`
```js
{
	"name": "mongo-mongoose",
	"version": "0.0.1",
	"scripts": {
		"dev": "env-cmd svelte-kit dev",
		"build": "svelte-kit build",
		"package": "svelte-kit package",
		"preview": "svelte-kit preview",
		"prepare": "svelte-kit sync",
		"lint": "prettier --ignore-path .gitignore --check --plugin-search-dir=. . && eslint --ignore-path .gitignore .",
		"format": "prettier --ignore-path .gitignore --write --plugin-search-dir=. ."
	},
	"devDependencies": {
		"@sveltejs/adapter-auto": "next",
		"@sveltejs/kit": "next",
		"env-cmd": "^10.1.0",
		"eslint": "^7.32.0",
		"eslint-config-prettier": "^8.3.0",
		"eslint-plugin-svelte3": "^3.2.1",
		"mongodb": "^4.3",
		"mongoose": "^6.2.5",
		"prettier": "^2.5.1",
		"prettier-plugin-svelte": "^2.5.0",
		"svelte": "^3.44.0"
	},
	"type": "module"
}
```

When I now run `npm list | grep "mongo"` I get to following output.
```shell
mongo-mongoose@0.0.1 /shared/httpd/mongo-mongoose
├─┬ mongodb@4.4.1
│ ├─┬ mongodb-connection-string-url@2.5.2
├─┬ mongoose@6.2.6
│ ├─┬ mongodb@4.3.1
│ │ ├── mongodb-connection-string-url@2.5.2 deduped
```

However I still get `MongooseError: Must call `setClient()` with an instance of MongoClient`.

**Reply by vkarpov15** [reply]
You're talking about the MongoDB server version, which is not the [mongodb npm package](https://www.npmjs.com/package/mongodb).

Remove "mongodb" from your `devDependencies` and it should work fine.

**Reply by robots4life** [reply]
```shell
1:38:49 PM [vite] Error when evaluating SSR module /node_modules/mongodb/lib/index.js?v=4335bc82:
ReferenceError: exports is not defined
    at eval (/node_modules/mongodb/lib/index.js?v=4335bc82:4:23)
    at instantiateModule (/shared/httpd/mongo-mongoose/node_modules/vite/dist/node/chunks/dep-9c153816.js:56193:15)
1:38:49 PM [vite] Error when evaluating SSR module /shared/httpd/mongo-mongoose/src/routes/index.js:
ReferenceError: exports is not defined
    at eval (/node_modules/mongodb/lib/index.js?v=4335bc82:4:23)
    at instantiateModule (/shared/httpd/mongo-mongoose/node_modules/vite/dist/node/chunks/dep-9c153816.js:56193:15)
exports is not defined
ReferenceError: exports is not defined
    at eval (/node_modules/mongodb/lib/index.js?v=4335bc82:4:23)
    at instantiateModule (/shared/httpd/mongo-mongoose/node_modules/vite/dist/node/chunks/dep-9c153816.js:56193:15)
```
    
How it is supposed to work by removing `mongodb` from `devDependencies` when I have

```js
import mongoose from 'mongoose';

import { MongoClient } from 'mongodb';
```

in https://github.com/robots4life/mongo-mongoose/blob/master/src/routes/index.js ?

If you find the time, would be cool to see this work in a fork of yours.

**Reply by vkarpov15** [reply]
`import { MongoClient } from 'mongodb'` _should_ still work because Mongoose pulls in mongodb.

But, if it doesn't, make sure the version of MongoDB in `devDependencies` is exactly equal to the one that your version of Mongoose pulls in. In this case, it should be 4.3.1. Because you're putting in `"mongodb": "^4.3"`, you're trying to use a `MongoClient` instance from `mongodb@4.4.1` with `mongoose@6.2`, and that version of Mongoose only supports `mongodb@6.2.1`

**Reply by robots4life** [reply]
https://github.com/robots4life/mongo-mongoose/blob/master/package.json

`"mongodb": "4.3.1",`
`"mongoose": "6.2.5",`

Ok, this works, at least it shows no errors. Kindly help me understand _why_ these version work together please. :pray: 

The MongoDB server version I use is 5.0.
When I look at https://mongoosejs.com/docs/compatibility.html it says `MongoDB Server 5.x: mongoose ^6.0.0`.

![Screenshot_20220420_142455](https://user-images.githubusercontent.com/34193243/164229823-85b3fb55-9d25-46a4-8a3f-4d1a6d80ac75.png)

So this means for MongoDB Server version 5.0 I need to use `mongoose higher than 6.0.0`, correct ?

When I look at https://www.mongodb.com/docs/drivers/node/current/compatibility/#mongodb-compatibility I see for MongoDB Server version 5.0 that the latest supported NodeJS driver version/npm package version is 4.5.

![Screenshot_20220420_142004](https://user-images.githubusercontent.com/34193243/164229030-e9f99829-cbf8-4ab4-8641-4b719502a588.png)


So this is currently exactly the version from here https://www.npmjs.com/package/mongodb. It says 4.5.0.

![Screenshot_20220420_142128](https://user-images.githubusercontent.com/34193243/164229274-96b97bf1-d454-495b-9b9a-1419de8e130c.png)

So.. if I wanted to use MongoDB Server version 5.0 and mongodb npm package version 4.5.0, _what_ version of mongoose npm package would I have to use with it please ?

edit:
If I have
```shell
"mongodb": "4.5.0",
"mongoose": "6.2.5",
```
it does not work.
However if I have
```shell
"mongodb": "4.5.0",
"mongoose": "6.3.0",
```
it does work.

Ok, this helps. The _very last thing_ I am now after is a **compatibility table between mongodb npm package and mongoose npm package versions**. Do you have that somewhere please ?

**Reply by vkarpov15** [reply]
@robots4life the ideal way is to check what version of MongoDB mongoose installs. So `cat node_modules/mongoose/package.json | grep "mongodb"`. Does that work for you or would you prefer a page in the docs?

**Reply by robots4life** [reply]
It would be great to have a page in the docs consolidating the info from this ticket or how to find compatible versions between mongodb and mongoose.

**Reply by Uzlopak** [reply]
mongodb is not a devDependency but a dependency. There is actually no need to add mongodb as a dependency next to mongoose in your project.

**Reply by robots4life** [reply]
@Uzlopak 
> There is actually no need to add mongodb as a dependency next to mongoose in your project.

Why ?

> This is useful if you already have a MongoClient instance, and want to reuse it.

Without including mongodb I cannot do https://github.com/robots4life/mongo-mongoose/blob/master/src/routes/index.js#L3 or https://github.com/robots4life/mongo-mongoose/blob/master/src/routes/index.js#L10.

**Reply by Uzlopak** [reply]
If you use mongoose as a dependency, then npm looks for its dependencies and installs them in the root of node_modules. So if you only use mongoose as a dependency it will put the mongodb package in the root of node_modules. If you use mongodb as a package next to mongoose, npm checks if they are matching versions. If they dont match, then npm installs the one specified in your package.json under node_modules/mongodb and the one from mongoose into node_modules/mongoose/node_modules/mongodb. 

If you use mongoose, which has mongodb as dependency, and another package, which has mongodb as a depenendency, then npm installs mongodb not on root level but in separate node_modules of each package, like mongoose/node_modules/mongodb and other_package/node_modules/mongodb.

This is also what you wrote in your comment. 

Actually: Why dont you let mongoose handle the connection? Why do you separately create the client and connect it to mongoose?

**Reply by robots4life** [reply]
@vkarpov15 What @Uzlopak wrote here https://github.com/Automattic/mongoose/discussions/11494#discussioncomment-2678491 would be good to have in the docs perhaps. For me this was not apparent.

@Uzlopak To your question, a) I don't always need mongoose and b) I wanted to learn how to get https://mongoosejs.com/docs/api/connection.html#connection_Connection-setClient to work in cases where I want to use mongoose while already having a mongodb connection.

Something else given your detailed explanation.
When I just do `npm install mongoose --save` in a new project, will I still have `mongodb` npm package available and still be able to do https://github.com/robots4life/mongo-mongoose/blob/master/src/routes/index.js#L3 without explicitly having installed `mongodb` ? Have I understood this correct from your comment ?

**Reply by Uzlopak** [reply]
Yes:

![Screenshot from 2022-05-03 10-32-03](https://user-images.githubusercontent.com/5059100/166424648-2e5cccdb-bab9-460c-b6fa-fb1ed2767824.png)
